antlr plugin's generated source files are not properly tracked for task wiring purposes

### Expected Behavior

Given a build script like below, running a task that consumes generated source files from antlr should not trigger any build failures due to failing to declare explicit task dependencies. Gradle's `SourceSets` concept should handle this transparently.



### Current Behavior

Running `./gradlew lib:generateGrammarSource lib:printFiles --console=verbose` will result in a build failure due to failure to declare explicit task dependencies (which should not be required).

```
./gradlew lib:generateGrammarSource lib:printFiles --console=verbose
> Task :lib:generateGrammarSource NO-SOURCE
> Task :lib:printFiles FAILED

FAILURE: Build failed with an exception.

* What went wrong:
A problem was found with the configuration of task ':lib:printFiles' (type 'MyTask').
  - Gradle detected a problem with the following location: '/Users/trobalik/Development/temp/antlr/lib/build/generated-src/antlr/main'.
    
    Reason: Task ':lib:printFiles' uses this output of task ':lib:generateGrammarSource' without declaring an explicit or implicit dependency. This can lead to incorrect results being produced, depending on what order the tasks are executed.
    
    Possible solutions:
      1. Declare task ':lib:generateGrammarSource' as an input of ':lib:printFiles'.
      2. Declare an explicit dependency on ':lib:generateGrammarSource' from ':lib:printFiles' using Task#dependsOn.
      3. Declare an explicit dependency on ':lib:generateGrammarSource' from ':lib:printFiles' using Task#mustRunAfter.

...
```

### Context (optional)

I ran into this while upgrading [my plugin](https://github.com/autonomousapps/dependency-analysis-android-gradle-plugin/tree/main) to Gradle 8.2.1 (from 7.5). I ended up [disabling one of my tests](https://github.com/autonomousapps/dependency-analysis-android-gradle-plugin/blob/main/src/functionalTest/groovy/com/autonomousapps/jvm/AntlrSpec.groovy#L12) so that I wouldn't be blocked. There is also an [open issue](https://github.com/autonomousapps/dependency-analysis-android-gradle-plugin/issues/685) on my plugin for [this deprecation warning](https://github.com/gradle/gradle/issues/18563) in Gradle 7.x.

### Steps to Reproduce

Take the following build script. Use Gradle 8.2.1.
```groovy
// lib/build.gradle (from `gradle init`)
plugins {
  id 'java-library'
  id 'antlr'
}

repositories {
  mavenCentral()
}

java {
  toolchain {
    languageVersion = JavaLanguageVersion.of(11)
  }
}

tasks.register('printFiles', MyTask) {
  sourceFiles.setFrom(sourceSets.main.allJava.sourceDirectories.asFileTree)
  output.set(layout.buildDirectory.file("output.txt"))
}

@CacheableTask
abstract class MyTask extends DefaultTask {

  @PathSensitive(PathSensitivity.RELATIVE)
  @InputFiles
  abstract ConfigurableFileCollection getSourceFiles()

  @OutputFile
  abstract RegularFileProperty getOutput()

  @TaskAction void action() {
    def output = output.get().asFile
    output.delete()

    def names = sourceFiles.collect {
      it.name
    }

    logger.quiet("Files: $names")

    output.write(names.toString())
  }
}
```
Now run
```
./gradlew lib:generateGrammarSource lib:printFiles
```

### Gradle version

8.2.1

### Build scan URL (optional)

https://scans.gradle.com/s/yqb63g6g5qal4/tests/task/:functionalTest/details/com.autonomousapps.jvm.AntlrSpec/antlr%20dep%20on%20antlr%20conf%20is%20not%20declared%20unused%20(Gradle%208.2.1)?top-execution=1

### Your Environment (optional)

_No response_

--- a/platforms/software/antlr/src/main/java/org/gradle/api/plugins/antlr/AntlrPlugin.java
+++ b/platforms/software/antlr/src/main/java/org/gradle/api/plugins/antlr/AntlrPlugin.java
@@ -19,7 +19,6 @@ package org.gradle.api.plugins.antlr;
 import org.gradle.api.Action;
 import org.gradle.api.Plugin;
 import org.gradle.api.Project;
-import org.gradle.api.Task;
 import org.gradle.api.artifacts.Configuration;
 import org.gradle.api.internal.project.ProjectInternal;
 import org.gradle.api.internal.tasks.DefaultSourceSet;
@@ -29,6 +28,7 @@ import org.gradle.api.plugins.JavaLibraryPlugin;
 import org.gradle.api.plugins.JavaPluginExtension;
 import org.gradle.api.plugins.antlr.internal.DefaultAntlrSourceDirectorySet;
 import org.gradle.api.tasks.SourceSet;
+import org.gradle.api.tasks.TaskProvider;
 
 import javax.inject.Inject;
 import java.io.File;
@@ -82,28 +82,23 @@ public abstract class AntlrPlugin implements Plugin<Project> {
                     //    naming conventions via call to sourceSet.getTaskName()
                     final String taskName = sourceSet.getTaskName("generate", "GrammarSource");
 
-                    // 3) Set up the Antlr output directory (adding to javac inputs!)
+                    // 3) Set up the Antlr output directory
                     final String outputDirectoryName = project.getBuildDir() + "/generated-src/antlr/" + sourceSet.getName();
                     final File outputDirectory = new File(outputDirectoryName);
-                    sourceSet.getJava().srcDir(outputDirectory);
 
-                    project.getTasks().register(taskName, AntlrTask.class, new Action<AntlrTask>() {
+                    // 4) Register a source-generating task, and
+                    TaskProvider<AntlrTask> antlrTask = project.getTasks().register(taskName, AntlrTask.class, new Action<AntlrTask>() {
                         @Override
                         public void execute(AntlrTask antlrTask) {
                             antlrTask.setDescription("Processes the " + sourceSet.getName() + " Antlr grammars.");
-                            // 4) set up convention mapping for default sources (allows user to not have to specify)
+                            // 4.1) set up convention mapping for default sources (allows user to not have to specify)
                             antlrTask.setSource(antlrSourceSet);
                             antlrTask.setOutputDirectory(outputDirectory);
                         }
                     });
 
-                    // 5) register fact that antlr should be run before compiling
-                    project.getTasks().named(sourceSet.getCompileJavaTaskName(), new Action<Task>() {
-                        @Override
-                        public void execute(Task task) {
-                            task.dependsOn(taskName);
-                        }
-                    });
+                    // 5) Add that task's outputs to the Java source set
+                    sourceSet.getJava().srcDir(antlrTask.map(t -> t.getOutputDirectory()));
                 }
             });
     }

--- a/platforms/software/antlr/src/main/java/org/gradle/api/plugins/antlr/AntlrPlugin.java
+++ b/platforms/software/antlr/src/main/java/org/gradle/api/plugins/antlr/AntlrPlugin.java
@@ -98,7 +98,7 @@ public abstract class AntlrPlugin implements Plugin<Project> {
                     });
 
                     // 5) Add that task's outputs to the Java source set
-                    sourceSet.getJava().srcDir(antlrTask.map(t -> t.getOutputDirectory()));
+                    sourceSet.getJava().srcDir(antlrTask);
                 }
             });
     }

--- a/platforms/software/antlr/src/test/groovy/org/gradle/api/plugins/antlr/AntlrPluginTest.groovy
+++ b/platforms/software/antlr/src/test/groovy/org/gradle/api/plugins/antlr/AntlrPluginTest.groovy
@@ -93,4 +93,19 @@ class AntlrPluginTest extends AbstractProjectBuilderSpec {
         def main = project.sourceSets.main
         main.extensions.extensionsSchema.find { it.name == 'antlr' }.publicType == typeOf(AntlrSourceDirectorySet)
     }
+
+    def 'adds task dependency to sourcesJar'() {
+        when:
+        project.pluginManager.apply(AntlrPlugin)
+
+        and:
+        project.java {
+            withSourcesJar()
+        }
+
+        then:
+        def generateGrammarSource = project.tasks.generateGrammarSource
+        def sourcesJar = project.tasks.sourcesJar
+        sourcesJar.taskDependencies.getDependencies(null).contains(generateGrammarSource)
+    }
 }

